File: frontend/src/components/benchmark/content/louvain/implementations/louvain.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <random>
#include <algorithm>

using namespace std;

class Louvain {
private:
    unordered_map<int, unordered_set<int>> adjacency;
    unordered_set<int> nodes;
    unordered_map<int, int> communities;
    map<pair<int, int>, double> weights;
    double m = 0.0;

    double degree(int node) {
        double deg = 0.0;
        for (int neighbor : adjacency[node]) {
            int u = node, v = neighbor;
            if (u > v) swap(u, v);
            deg += weights[{u, v}];
        }
        return deg;
    }

    double weight(int u, int v) {
        if (u > v) swap(u, v);
        return weights[{u, v}];
    }

    double community_degree(int node, int community) {
        double deg = 0.0;
        for (int neighbor : adjacency[node]) {
            if (communities[neighbor] == community) {
                deg += weight(node, neighbor);
            }
        }
        return deg;
    }

    double total_degree(int community) {
        double total = 0.0;
        for (int node : nodes) {
            if (communities[node] == community) {
                total += degree(node);
            }
        }
        return total;
    }

    double modularity_gain(int node, int community) {
        double ki = degree(node);
        double kic = community_degree(node, community);
        double total = total_degree(community);
        return (kic - (total * ki) / (2 * m)) / m;
    }

    void initialize_communities() {
        for (int node : nodes) {
            communities[node] = node;
        }
    }

    bool first_phase() {
        bool changed = false;
        vector<int> node_list(nodes.begin(), nodes.end());
        shuffle(node_list.begin(), node_list.end(), default_random_engine(random_device{}()));
        
        for (int node : node_list) {
            int best_community = communities[node];
            double best_gain = 0.0;
            
            for (int neighbor : adjacency[node]) {
                int community = communities[neighbor];
                if (community == communities[node]) continue;
                double gain = modularity_gain(node, community);
                if (gain > best_gain) {
                    best_gain = gain;
                    best_community = community;
                }
            }
            
            if (best_community != communities[node]) {
                communities[node] = best_community;
                changed = true;
            }
        }
        
        return changed;
    }

    void second_phase() {
        unordered_map<int, unordered_set<int>> new_adjacency;
        unordered_set<int> new_nodes;
        map<pair<int, int>, double> new_weights;
        double new_m = 0.0;
        
        unordered_map<int, int> community_map;
        int next_id = 0;
        for (auto& pair : communities) {
            if (community_map.find(pair.second) == community_map.end()) {
                community_map[pair.second] = next_id++;
            }
        }
        
        for (auto& pair : communities) {
            int new_community = community_map[pair.second];
            if (new_adjacency.find(new_community) == new_adjacency.end()) {
                new_adjacency[new_community] = unordered_set<int>();
            }
            new_nodes.insert(new_community);
        }
        
        for (auto& pair : weights) {
            int u = pair.first.first;
            int v = pair.first.second;
            double w = pair.second;
            
            int cu = community_map[communities[u]];
            int cv = community_map[communities[v]];
            
            if (cu == cv) {
                new_m += w;
                continue;
            }
            
            if (cu > cv) swap(cu, cv);
            new_weights[{cu, cv}] += w;
            new_adjacency[cu].insert(cv);
            new_adjacency[cv].insert(cu);
            new_m += w;
        }
        
        for (int community : new_nodes) {
            if (new_adjacency.find(community) == new_adjacency.end()) {
                new_adjacency[community] = unordered_set<int>();
            }
        }
        
        adjacency = new_adjacency;
        nodes = new_nodes;
        weights = new_weights;
        m = new_m;
        
        unordered_map<int, int> new_communities;
        for (auto& pair : communities) {
            new_communities[pair.first] = community_map[pair.second];
        }
        communities = new_communities;
    }

public:
    void add_edge(int u, int v, double weight = 1.0) {
        adjacency[u].insert(v);
        adjacency[v].insert(u);
        nodes.insert(u);
        nodes.insert(v);
        if (u > v) swap(u, v);
        weights[{u, v}] += weight;
        m += weight;
    }

    void build_from_edges(const vector<pair<int, int>>& edges) {
        for (auto& edge : edges) {
            add_edge(edge.first, edge.second);
        }
    }

    unordered_map<int, int> detect_communities() {
        initialize_communities();
        
        while (true) {
            bool improved = false;
            while (first_phase()) {
                improved = true;
            }
            if (!improved) break;
            second_phase();
        }
        
        return communities;
    }
};

File: frontend/src/components/benchmark/content/louvain/implementations/louvain.go
package main

import (
    "math/rand"
    "time"
)

type Louvain struct {
    adjacency  map[int]map[int]bool
    nodes      map[int]bool
    communities map[int]int
    weights    map[[2]int]float64
    m          float64
}

func NewLouvain() *Louvain {
    return &Louvain{
        adjacency:  make(map[int]map[int]bool),
        nodes:      make(map[int]bool),
        communities: make(map[int]int),
        weights:    make(map[[2]int]float64),
        m:          0.0,
    }
}

func (l *Louvain) AddEdge(u, v int, weight float64) {
    if l.adjacency[u] == nil {
        l.adjacency[u] = make(map[int]bool)
    }
    if l.adjacency[v] == nil {
        l.adjacency[v] = make(map[int]bool)
    }

    l.adjacency[u][v] = true
    l.adjacency[v][u] = true

    l.nodes[u] = true
    l.nodes[v] = true

    if u > v {
        u, v = v, u
    }
    l.weights[[2]int{u, v}] += weight
    l.m += weight
}

func (l *Louvain) BuildFromEdges(edges [][2]int) {
    for _, edge := range edges {
        l.AddEdge(edge[0], edge[1], 1.0)
    }
}

func (l *Louvain) degree(node int) float64 {
    deg := 0.0
    for neighbor := range l.adjacency[node] {
        u, v := node, neighbor
        if u > v {
            u, v = v, u
        }
        deg += l.weights[[2]int{u, v}]
    }
    return deg
}

func (l *Louvain) weight(u, v int) float64 {
    if u > v {
        u, v = v, u
    }
    return l.weights[[2]int{u, v}]
}

func (l *Louvain) communityDegree(node, community int) float64 {
    deg := 0.0
    for neighbor := range l.adjacency[node] {
        if l.communities[neighbor] == community {
            deg += l.weight(node, neighbor)
        }
    }
    return deg
}

func (l *Louvain) totalDegree(community int) float64 {
    total := 0.0
    for node := range l.nodes {
        if l.communities[node] == community {
            total += l.degree(node)
        }
    }
    return total
}

func (l *Louvain) modularityGain(node, community int) float64 {
    ki := l.degree(node)
    kic := l.communityDegree(node, community)
    total := l.totalDegree(community)
    return (kic - (total * ki) / (2 * l.m)) / l.m
}

func (l *Louvain) initializeCommunities() {
    for node := range l.nodes {
        l.communities[node] = node
    }
}

func (l *Louvain) firstPhase() bool {
    changed := false
    nodeList := make([]int, 0, len(l.nodes))
    for node := range l.nodes {
        nodeList = append(nodeList, node)
    }

    rand.Seed(time.Now().UnixNano())
    rand.Shuffle(len(nodeList), func(i, j int) {
        nodeList[i], nodeList[j] = nodeList[j], nodeList[i]
    })

    for _, node := range nodeList {
        bestCommunity := l.communities[node]
        bestGain := 0.0

        for neighbor := range l.adjacency[node] {
            community := l.communities[neighbor]
            if community == l.communities[node] {
                continue
            }
            gain := l.modularityGain(node, community)
            if gain > bestGain {
                bestGain = gain
                bestCommunity = community
            }
        }

        if bestCommunity != l.communities[node] {
            l.communities[node] = bestCommunity
            changed = true
        }
    }

    return changed
}

func (l *Louvain) secondPhase() {
    newAdjacency := make(map[int]map[int]bool)
    newNodes := make(map[int]bool)
    newWeights := make(map[[2]int]float64)
    newM := 0.0

    communityMap := make(map[int]int)
    nextId := 0
    communitySet := make(map[int]bool)
    for _, community := range l.communities {
        if !communitySet[community] {
            communitySet[community] = true
            communityMap[community] = nextId
            nextId++
        }
    }

    for node, community := range l.communities {
        newCommunity := communityMap[community]
        if newAdjacency[newCommunity] == nil {
            newAdjacency[newCommunity] = make(map[int]bool)
        }
        newNodes[newCommunity] = true
    }

    for edge, weight := range l.weights {
        u := edge[0]
        v := edge[1]
        cu := communityMap[l.communities[u]]
        cv := communityMap[l.communities[v]]

        if cu == cv {
            newM += weight
            continue
        }

        if cu > cv {
            cu, cv = cv, cu
        }
        newWeights[[2]int{cu, cv}] += weight
        if newAdjacency[cu] == nil {
            newAdjacency[cu] = make(map[int]bool)
        }
        if newAdjacency[cv] == nil {
            newAdjacency[cv] = make(map[int]bool)
        }
        newAdjacency[cu][cv] = true
        newAdjacency[cv][cu] = true
        newM += weight
    }

    for community := range newNodes {
        if newAdjacency[community] == nil {
            newAdjacency[community] = make(map[int]bool)
        }
    }

    l.adjacency = newAdjacency
    l.nodes = newNodes
    l.weights = newWeights
    l.m = newM

    newCommunities := make(map[int]int)
    for node, community := range l.communities {
        newCommunities[node] = communityMap[community]
    }
    l.communities = newCommunities
}

func (l *Louvain) DetectCommunities() map[int]int {
    l.initializeCommunities()

    for {
        improved := false
        for l.firstPhase() {
            improved = true
        }
        if !improved {
            break
        }
        l.secondPhase()
    }

    return l.communities
}